src/TestTrait.php

```diff
@@ -126,48 +169,38 @@ protected function shrinkingTimeLimit($shrinkingTimeLimit)
     }
 
     /**
+     * @param string|RandomRande $randFunction mt_rand, rand or a RandomRange
      * @return self
      */
     protected function withRand($randFunction)
     {
-        // TODO: invert and wrap rand, srand into objects?
-        if ($randFunction instanceof \Eris\Random\RandomRange) {
-            $this->randFunction = function ($lower = null, $upper = null) use ($randFunction) {
-                return $randFunction->rand($lower, $upper);
-            };
-            $this->seedFunction = function ($seed) use ($randFunction) {
-                return $randFunction->seed($seed);
-            };
+        if ($randFunction === 'mt_rand') {
+            $this->randFunction = new RandomRange(new MtRandSource());
+            return $this;
         }
-        if (is_callable($randFunction)) {
-            switch ($randFunction) {
-                case 'rand':
-                    $seedFunction = 'srand';
-                    break;
-                case 'mt_rand':
-                    $seedFunction = 'mt_srand';
-                    break;
-                default:
-                    throw new BadMethodCallException("When specifying random generators different from the standard ones, you must also pass a \$seedFunction callable that will be called to seed it.");
-            }
+        if ($randFunction === 'rand') {
+            $this->randFunction = new RandomRange(new RandSource());
+            return $this;
+        }
+        if ($randFunction instanceof RandomRange) {
             $this->randFunction = $randFunction;
-            $this->seedFunction = $seedFunction;
+            return $this;
         }
-        return $this;
+        throw new BadMethodCallException("When specifying random generators different from the standard ones, you must also pass a \$seedFunction callable that will be called to seed it.");
     }
 
     /**
      * forAll($generator1, $generator2, ...)
-     * @return Quantifier\ForAll
+     * @return ForAll
      */
     public function forAll()
     {
-        call_user_func($this->seedFunction, $this->seed);
+        $this->randFunction->seed($this->seed);
```

Comment:
`$this->randFunction` should be renamed to `$this->randomRange` now to reflect its new type. I'm not sure about the implications, but I wouldn't expect the field to be used directly by client code